#3 로그인 기능

0hun opened the pull request.

0hun committed:
#3 로그인 기능

0hun committed:
Update README.md

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -53,4 +59,45 @@ public void initBinder(WebDataBinder webDataBinder) {
+     * @param httpSession
+     * @return ResponseEntity
+     */
+    @GetMapping("/login")
```
RequestBody를 이용하기 때문에 `POST`로 해주는게 좋을 것 같습니다~
`GET`은 일반적으로 request body를 사용하지 않습니다

0hun replied:
넵 post로 변경하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -53,4 +59,45 @@ public void initBinder(WebDataBinder webDataBinder) {
+        }
+        else if (Account.Status.DEFAULT.equals(account.getStatus())) {
+            httpSession.setAttribute("account", account.getId());
+            responseEntity = new ResponseEntity(account.getId(), HttpStatus.OK);
```
클라이언트에서 userId는 이미 갖고있기 때문에 response에 id를 굳이 담아줘야할지 고민이 필요할 것 같고요~
만약 담아주지 않아도 된다면 `new`로 계속 새로운 객체를 생성하기보다는 상수를 사용하면 불필요한 객체생성을 줄일 수 있습니다.
객체 생성을 줄이면 왜 성능이 좋아지는지도 주석으로 한번 달아주시면 좋겠네요~

0hun replied:
넵 주석달도록 하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -53,4 +59,45 @@ public void initBinder(WebDataBinder webDataBinder) {
+    /**
+     * 회원 로그아웃 메서드.
+     *
+     * @author jun
```
혹시나 해서 말씀드리면 다른 소스를 참고하거나 베끼는건 최대한 지양하셔야합니다.
개발방향을 파악하는데에 참고하는건 괜찮지만 특히 복붙을 제일 조심하셔야하는데요. 만나서 말씀드린것처럼 복붙은 코드 한줄한줄의 의미를 깊게 파악하는데에 엄청난 방해가됩니다~

이 프로젝트는 저에게 확인받기 위한 프로젝트가 아니라 본인 개발실력을 늘리려는 목적이므로 단순히 기능완성을 목표로 하기보다는 평소에 만들어봤던 기능 하나를 다시 구현하는 것이더라도 한줄한줄 최대한 깊게 고민하면서 진행하는게 중요합니다.
평소에 그냥 지나치던 개념들이 많은 트래픽이 들어왔을 때, 많은 인원들이 같이 작업할때를 고려해보면 모두 중요한 개념들이기 때문입니다

0hun replied:
넵 유의하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -53,4 +59,45 @@ public void initBinder(WebDataBinder webDataBinder) {
+     *
+     * @author jun
+     * @param session 현제 접속한 세션
+     * @return 로그인 하지 않았을 시 401코드를 반환하고 result:NO_LOGIN 반환 로그아웃 성공시 200 코드를 반환
```
왜 401 코드를 반환할까요?

0hun replied:
로그인이 안됐을경우 해당 리소스에 대해 인증권한이 없어 401코드를 반환한다는 의미로 사용했습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -53,4 +59,45 @@ public void initBinder(WebDataBinder webDataBinder) {
+     */
+    @GetMapping("logout")
+    public void logout(HttpSession session) {
+        accountService.logout(session);
```
왜 logout 로직을 accountService에서 담당하게 하셨나요?
`Controller`, `Service`, `DAO`의 역할을 어떻게 구분했는지 클래스 주석에 정리해주시면 좋을 것 같습니다~

0hun replied:
넵넵 주석에 정리하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/dto/LoginDto.java:
```diff
@@ -0,0 +1,19 @@
+import javax.validation.constraints.NotBlank;
+
+@Getter
+@Setter
```
이 DTO는 값을 중간에 수정할 필요가 있을까요?
`java immutable`에 대해 검색해보시면 좋을 것 같습니다

0hun replied:
login시에만 필요하므로 이후 수정할 필요가 없을것 같습니다. java immutable에 대해 확인해보도록 하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/dto/LoginDto.java:
```diff
@@ -0,0 +1,19 @@
+
+@Getter
+@Setter
+@Builder
```
여기에 Builder를 사용하신 이유가 무엇인가요?

0hun replied:
loginDto 객체 매개변수에 값을 넣을때 setter를 쓰면 set method로 코드량이 늘어나서 줄이기 위해
사용하였습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -30,4 +32,13 @@ public boolean existsById(String id) {
     }
+
+    public Account login(String accountId, String password) {
+        Account account = accountMapper.findByIdAndPassword(accountId, password);
```
이거는 그냥 한 줄로 줄일 수 있지 않을까요?
`return accountMapper.findByIdAndPassword(accountId, password);`

0hun replied:
넵 수정 반영하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -30,4 +32,13 @@ public boolean existsById(String id) {
+    }
+
+    public void logout(HttpSession session) {
+        session.removeAttribute("account");
```
`"account"`라는 값을 하드코딩하기보다는 상수로 만들어놓고 돌려쓰는것이 유지보수성에 좋습니다.
만약 키를 바꾸게 된다면 모든 소스를 다 뒤져가면서 이 값을 바꿔야할 가능성이 생기기 때문입니다

0hun replied:
session을 저장할 상수라면 어떻게 만들어야할지 감이 안오는데, enum 형태로 만들어서 사용하면될까요?

f-lab-dev replied:
`SessionKeys` 이런 상수를 담는 클래스를 따로 만드셔도 되고, 혹은 properties로 분리해도 됩니다~

f-lab-dev commented on src/main/resources/mapper/account.xml:
```diff
@@ -21,6 +21,13 @@
     </select>
 
+    <select id="findByIdAndPassword" resultMap="accountResultMap">
+        SELECT *
```
어떤 값을 사용할지 필요한 컬럼들을 하나하나 명시해주는게 좋을 것 같네요~

0hun replied:
넵 수정 반영하겠습니다.

f-lab-dev commented on src/test/java/com/younghun/hibusgo/controller/AccountControllerTest.java:
```diff
@@ -85,4 +87,42 @@ void addAccount_with_correct_input() throws Exception {
+    @DisplayName("로그인")
+    @Test
+    void account_login() throws Exception {
+//        AccountDto accountDto = AccountDto.builder()
```
이런 주석같은 로컬 테스트용 코드는 커밋하지 않는게 좋습니다~

0hun replied:
넵 수정하겠습니다.

f-lab-dev commented on src/test/java/com/younghun/hibusgo/controller/AccountControllerTest.java:
```diff
@@ -85,4 +87,42 @@ void addAccount_with_correct_input() throws Exception {
+
+    @DisplayName("로그인")
+    @Test
+    void account_login() throws Exception {
```
이 테스트는 단위테스트인가요 통합테스트인가요?
어떤 것인지 차이를 비교해보고 주석으로 달아주시면 좋을 것 같습니다

0hun replied:
통합테스트라고 생각해서 추가했는데, 차이점에 대해 찾아보도록 하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/dto/AccountDto.java:
```diff
@@ -29,7 +29,7 @@
     private String password;
 
-    // 이
+    // 이름
```
이미 name이라는 변수명에서 이름이라는 의미를 유추할 수 있으므로 이 주석은 의미가 없는 것 같습니다~

0hun replied:
넵 수정 반영하겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/AccountController.java:
```diff
@@ -11,8 +12,13 @@
 import org.springframework.validation.Errors;
 import org.springframework.web.bind.WebDataBinder;
 import org.springframework.web.bind.annotation.*;
```
이런 `*` import는 컨벤션에서 추천하지는 않는 방식입니다.
`intellij checkstyle`로 검색해보시고 괜찮은 코드 컨벤션을 찾아보시는게 좋을 것 같네요~

0hun replied:
넵 확인보겠습니다.

0hun committed:
#3 로그인 기능

리뷰 내용 반영

0hun committed:
#3 로그인 기능

로그인 컨트롤러 생성하여 로그인, 로그아웃 로직 분리

0hun committed:
Merge branch 'issue/2' into issue/3

0hun committed:
#3 로그인 기능

리뷰 내용 반영
seesion key을 상수값으로 저장하도록 수정

0hun committed:
#3 로그인 기능

리뷰 내용 반영
seesion key을 상수값으로 저장하도록 수정

0hun committed:
#3 로그인 기능

리뷰 내용 반영
seesion key를 enum으로 변경

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+/**
+ * Controller : 사용자의 요청을 어떻게 처리 결정하는 역할
+ *
+ * Service: Controller에서 넘어온 요청을 처리하는 역할
```
Service는 단순히 Controller와 DAO를 연결지어주는 역할이 아닙니다~
좀 더 자세히 알아보시면 좋을 것 같습니다

0hun replied:
넵 조사해보겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+ * Service: Controller에서 넘어온 요청을 처리하는 역할
+ * service method는 한번에 하나의 비지니스로직만 수행
+ *
+ * DAO: Service에서 넘어온 요청 값을 기준으로 database 조회
```
`DAO`는 데이터베이스만 조회하는게 아닙니다.
DAO의 뜻을 알아보시고 어떤 계층을 나타내는 건지 명확히 써주시면 좋을 것 같네요

0hun replied:
넵 조사해보겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+
+
+/**
+ * Controller : 사용자의 요청을 어떻게 처리 결정하는 역할
```
세 계층에 대해 더 자세히 알아보시면 좋을 것 같습니다. Controller는 어떤 역할을 맡는 계층인지 명확하게 이해하고 있어야합니다~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+     *  new ResponseEntity() 새로운 객체를 생성하여 return 하는 대신 ResponseEntity에 상수를 담아 return
+     *  객체를 new로 생성을 많이 하면 다중의 사용자가 접근했을 때 새로운 객체가 자주 생성되고, GC가 자주 발생한다.
+     *  GC를 수행하는 동안 많은 메모리 소비와 jvm 수행이 멈추고 요청을 처리하는 동안 대기하는 경우가 발생한다.
+     *  이러한 이유로 새로운 객체롤 최대한 적게 생성하도록 지향.
```
프로젝트보다 책의 비중을 더 높혀야하는 이유가 여기에도 있습니다. 추천드린 책을 계속 읽다보면 GC가 나올거고 JVM이 언제 멈추는지 등 더 자세히 알고있어야하기 때문입니다

f-lab-dev commented on src/test/java/com/younghun/hibusgo/controller/AccountControllerTest.java:
```diff
@@ -26,6 +25,8 @@
  * @AutoConfigureMockMvc - Mock 테스트시 필요한 의존성을 제공. Service에서 호출하는 Bean을 주입해준다. 간단히 컨트롤러 클래스만 테스트 하고
  * 싶다면 Mockup Test를 사용할 수 있는데 service 객체에 @MockBean 어노테이션을 적용하는 것으로 이 어노테이션을 대체할 수 있다.
  *
+ *  단위 테스트 : 개발 단계에서 각 모듈(기능)이 개발 완료되는 시점에서 진행하는 테스트
```
단위테스트는 어떻게 개발해야할까요? 단위테스트는 개발완료되는 시점에만 진행하는걸까요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/domain/SessionKeys.java:
```diff
@@ -1,7 +1,12 @@
 package com.younghun.hibusgo.domain;
 
-public class SessionKeys {
+public enum SessionKeys {
```
이거를 굳이 enum으로 관리해야할까요? 이 값을 enum으로 만듦으로서 오는 이점이 뭐가 있을까요?
아래를 보면 enum으로 만들었기 때문에 불필요한 `String.valueOf()`를 사용하고 있는 것으로 보이네요~

0hun replied:
enum을 사용했을때 장점으로는 코드량이 줄어들고 컴파일 시점에 타입에 안정적이라는 장점이 있습니다.
값이 추후 늘어날수있겠지만, 현재 값이 적고 키값으로 넣어주려고 할때 불필요한 형변환을 하기때문에 
static fianl 상수로 변경하였습니다.

f-lab-dev replied:
오타 등을 컴파일 타임에 잡아낼 수 있다는건 상수도 같지 않을까요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
+        }
+        else if (Account.Status.DEFAULT.equals(account.getStatus())) {
+            ResponseEntity.status(HttpStatus.OK);
```
이 구문은 의미가 없어보입니다~

0hun replied:
로그인이 성공하였을때, return ResponeEntity.ok해서 필요없는건가요?

f-lab-dev replied:
네 맞습니다~ 이거는 값을 생성하는거지 전역적으로 설정하는건 아닙니다.
`Spring ResponseEntity Documentation`으로 검색하셔서 찾아보시면 어떤 클래스인지 알 수 있을겁니다

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -0,0 +1,85 @@
+        Account account = accountService.findByIdAndPassword(accountId, password);
+
+        if (account == null) {
+            return ResponseEntity.status(HttpStatus.UNAUTHORIZED).build();
```
로그인을 하지 않았기 때문에 로그인을 시도하는건데 **인증되지 않았다**는 뜻의 `UNAUTHORIZED`가 리턴되면 모양이 이상하지 않을까요?

0hun replied:
로그인을 시도하려고 했을때, 해당 유저가 db에 없어서 권한이 없다는 의미로 사용하였는데
어떤 걸로 사용해야 좋을까요

f-lab-dev reviewed (commented):
리뷰반영 커밋은 최대한 한곳에 모아주시면 좋을 것 같네요~
커밋 메세지에 변경내용을 적어주시고요

0hun committed:
#3 로그인 기능

리뷰 내용 반영

seesion keys enum 대신 static final 상수로 세션 키 생성

login 성공시 http ok code 설정하는 부분 수정
Login 실패시 status code를 204 code return

logout 실패시 403 code return

0hun committed:
#3 로그인 기능

logoutAccount가 존재 할 경우 비교하지 않고 바로 logout 로직 수행

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -16,20 +16,26 @@
 
 /**
- * Controller : 사용자의 요청을 어떻게 처리 결정하는 역할
+ * Controller(controller layer) : 비지니스 로직 처리 흐름 제어 및 데이터 변환 및 연산
```
비즈니스로직은 Service의 역할입니다~ 각 레이어가 어떤 역할을 하는지와 이렇게 레이어를 분리했을때의 이점도 생각해보시면 좋겠습니다

0hun replied:
controller가 serivce의 비지니스 로직 처리하는 흐름을 제어한다는 의미로 사용했습니다,

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -16,20 +16,26 @@
- * Service: Controller에서 넘어온 요청을 처리하는 역할
- * service method는 한번에 하나의 비지니스로직만 수행
+ * Service(Business Layer): controller를 통해 전달 받은 클라이언트의 요청을 처리하여 비지니스 로직 수행
+ * contoller layer와 persistence layer 연결, 다른 계층과 통신하기 위해 인터페이스 제공,
```
여기도 결국은 **Controller와 DAO의 중간자를 맡는다** 라는 말로 보이는데요
`Business Layer`라고 써두셨으니 한번 이 이름이 왜 붙었는지도 고민해보시면 좋겠네요~

0hun replied:
넵 알겠습니다.

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -1,7 +1,8 @@
 package com.younghun.hibusgo.service;
 
+import static com.younghun.hibusgo.controller.LoginController.ACCOUNT_MEMBER_ID;
```
service 레이어에서 controller 레이어 값을 참조하고 있네요.
이러면 레이어의 양 방향에 `순환 참조`가 일어나는건 아닐까요?

0hun replied:
순환참조에 대해선 생각을 못했는데, controller -> service -> dao 한 방향으로만 참조헤야된다는 말씀이신거죠?

f-lab-dev replied:
네 맞습니다~ 레이어간에는 순환참조가 일어나면 안됩니다.
그러면 "왜" 일어나면 안돼는 걸까요? 이렇게 레이어를 분리했을 때 이러한 관점에서 보면 어떤 이점이 있을까요?

0hun replied:
ACCOUNT_MEMBER_ID의 경우 static 변수라서 class를 로드하는 순간 static영역에 생성되어 순환참조문제가 발생하지 않았지만, 인스턴스 변수의 경우에는 순환참조가 발생합니다.
layer간 순환참조는 layer 간의 경계를 불분명하게 만들고 layer간 상호 의존은 재사용성을 떨어뜨리고 확장성에도 좋지 않다고 생각합니다.
layer간 한 방향으로 참조하게 하는것은 각 layer의 역할을 명확하게 하고 경계를 명확히 하게 합니다.
또한 재사용성과 확장성의 이점이 있습니다.

f-lab-dev commented on src/test/java/com/younghun/hibusgo/controller/AccountControllerTest.java:
```diff
@@ -22,11 +22,20 @@
+ *  단위 테스트 작성 방법
+ *  1. 버그를 찾는다.
+ *  2. 버그를 고쳤을 때 통과 할 만한 테스트 작성
+ *  3. 테스트를 통과할 때 까지 코드를 수정
```
이거도 추후에 테스트 관련한 책을 읽을건데요. 일단은 다른 공부를 위해 이정도로 마무리해두는게 좋겠네요

0hun replied:
넵 알겠습니다.

0hun committed:
#3 로그인 기능

리뷰 반영

ACCOUNT_MEMBER_ID 상수를 사용할때 service가 controller를
순환 참조 하지 않도록 상수를 service의 상수로 변경

0hun committed:
#3 로그인 기능

리뷰 반영

3 layer 주석 내용 수정

0hun committed:
#3 로그인 기능

리뷰 반영

3 layer 주석 내용 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/AccountService.java:
```diff
@@ -16,6 +15,7 @@
 public class AccountService {
 
     private final AccountMapper accountMapper;
+    public static final String ACCOUNT_MEMBER_ID = "ACCOUNT_MEMBER_ID";
```
1. 이것도 다른 관점에서 생각해보면 세션을 꼭 Account 쪽에서만 사용하게 될까요??
2. Service 레이어에서 세션의 존재에 대해 알고있어야할까요?
3. 세션은 어느 레이어에 종속적일까요?

0hun replied:
1. 세션은 account에서만 종속된 개념이 아니므로 다른 도메인에서도 사용할수있을것 같습니다.
2. service layer에서는 비지니스 로직을 처리하는 layer이므로 세션의 존재에대해 알 필요가 없다고 생각합니다.
3. 서블릿 컨테이너가 httpsession을 이용해 세션을 생성하고 관리하므로, presentaion layer에 종속적입니다. pesentation layer안에는 controller가 포함되어 있으므로 세션처리는 controller에서 처리하는게 계층적 관점으로 맞다고 생각합니다.

f-lab-dev replied:
3번 설명이 정말 좋네요~ 1번은 그러면 세션키 등 세션과 관련된 로직을 관리하는 클래스를 분리해주면 좋지 않을까요?

0hun replied:
일단 세션과 관련된 로직을 컨트롤러에서 처리하는데, 1번과 같은 이유에서 다른 클래스로 분류하는것이 좋을까요?

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -17,20 +17,19 @@
  * 3 계층 아키텍쳐 장점
  *  - 각 계층에 맞는 기능에 집중한 설계 및 개발
- *  - 각 계층별로 역할을 분담하여 동시 다발적인 개발
+ *  - 각 계층별로 역할을 분담하여 동시 다발적인 개
```
여기는 오타같네요~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -17,20 +17,19 @@
  *
  * DAO(Persistence Layer) : Business Layer에 전달 받은 데이터를 저장, 수정 ,삭제 하는 계층
- * db에서 조회한 데이터를 Business Layer에 응답 하기 위해 객체화한다.
+ * db에서 조회한 데이터를 Business Layer에 응답 하기 위해 객체화.
```
단순히 조회한 데이터를 객체화 하는 역할만 있을까요?
객체지향적인 관점에서 고민해보시면 좋을 것 같네요~

f-lab-dev reviewed (commented):
이제 주석 고치는 것 외에는 세션 키 관리 정도만 남았네요~
주석은 천천히 고치셔도 되고, 이것만 우선적으로 고쳐주시면 한번 다시 보고 머지해도 될 것 같습니다~

0hun committed:
#3 로그인 기능

리뷰 반영

3 layer 주석 내용 수정, session 처리를 LoginController에서 처리하도록 수정

0hun committed:
#3 로그인 기능

리뷰 반영

3 layer 주석 내용 수정

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -23,9 +23,11 @@
+ * DAO(Data Access Layer) : Business Layer에 통 전달 받은 데이터를 저장, 수정 ,삭제 하는 계층
+ * db에서 조회한 데이터를 Business Layer에 응답 하기 위해 객체화
+ * db에 관련된 작업을 수행하기 위해 필요한 쿼리문을 관리
+ * database와 접속을 관리하며, 개방 폐쇄 원칙을 적용하여 데이터베이스의 종류가 바뀌어도
```
말 그대로 데이터를 접근하는 계층을 추상화한 것이므로 `데이터베이스의 종류가 바뀌어도` 데이터 계층만 갈아끼면 된다는게 장점이죠~
잘 찾아보셨습니다 👍

f-lab-dev commented on src/main/java/com/younghun/hibusgo/controller/LoginController.java:
```diff
@@ -64,7 +67,7 @@ public ResponseEntity login(@RequestBody @NotNull LoginDto loginDto, HttpSession
         }
 
-        accountService.login(account.getId(), httpSession);
+        httpSession.setAttribute(ACCOUNT_MEMBER_ID, account.getId());
```
Service에서 처리하던 것을 Controller로 옮기셨는데요~
세션처리를 여기서 하는 이유를 레이어의 역할과 관련하여 주석으로 작성해주시면 좋을 것 같습니다~

0hun replied:
넵 주석 반영하겠습니다.

0hun committed:
#3 로그인 기능

리뷰 반영

주석 수정

0hun committed:
#3 로그인 기능

리뷰 반영

세션 키등 세션과 관련된 로직를 관리하는 클래스를 분리 생성

0hun committed:
#3 로그인 기능

리뷰 반영

SessionUtil 주석 추가

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/SessionUtil.java:
```diff
@@ -0,0 +1,17 @@
+
+import javax.servlet.http.HttpSession;
+
+public class SessionUtil {
```
이거는 `SessionService` 클래스로 분리하면 어떨까요?
인터페이스로 추상화한다면 내부 구현은 감춘채로 내부에서는 세션을 활용하여 로그인을 처리할 수 있게됩니다~

인터페이스로 추상화한다면 이런 기능들이 들어갈 듯 하네요. 추상화를 한번 잘 설계해보시면 좋을 듯 합니다~
- 로그인
- 로그아웃
- 로그인했는지 여부
- 로그인된 사용자의 정보 가져오기

0hun replied:
SessionService로 추상화해서 구현을 SessionUtil에서 하란 의미 인가요?

0hun committed:
#3 로그인 기능

리뷰 반영

SessionService 추상화 인터페이스 추가

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/SessionService.java:
```diff
@@ -0,0 +1,15 @@
+
+public interface SessionService {
+
+  void accountLogin(HttpSession session, String accountId);
```
메소드 인자로 HttpSession을 받으면 서비스 레이어에 세션에 대한 의존성이 들어갑니다.
메소드 인자가 아니더라도 스프링으로부터 세션을 주입받을 수 있습니다~ 한번 해보시는게 좋을 것 같네요.
이 메소드는 어떤걸 추상화 할것인가 -> `아이디와 비밀번호를 받아서 로그인처리를 한다` 이런 관점으로 접근해보시고요~

0hun replied:
로그인처리를 하기전에 아이디와 비밀번호를 받아서 계정이 있는지 체크하는 로직이 있는데, 비밀번호는 받지 않아도 되지 않나요?

f-lab-dev replied:
아 그럼 필요없겠군요~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/service/SessionService.java:
```diff
@@ -0,0 +1,15 @@
+
+import javax.servlet.http.HttpSession;
+
+public interface SessionService {
```
이름도 `AccountService`가 낫지 않을까요? Session이라는 이름은 세션방식의 로그인에 종속적이니까요

0hun replied:
AccountService가 이미 있어서 Login 처리에 관련된 로직을 처리할수있다는 의미에서 LoginService로 변경하였는데 괜찮을까요

f-lab-dev replied:
네 이름 적절해보입니다~

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/SessionUtil.java:
```diff
@@ -21,8 +25,29 @@ public static void accountLogin(HttpSession session, String accountId) {
+   */
+  @Override
+  public String getLoginAccountId(HttpSession session) {
+    return  (String) session.getAttribute(ACCOUNT_MEMBER_ID);
```
`session.getAttribute(ACCOUNT_MEMBER_ID)` 의 결과가 null이면 어떻게 될까요?

0hun replied:
아 넵 해당 세션으로된 키값이 없으면 exception이 발생하겠네요.

0hun committed:
#3 로그인 기능

리뷰 반영

세션 처리에 종속적인 의미가 될수 있어서 SessionService를 LoginService로 변경
메소드 인자로 httpsesion을 받지 않고 스프링으로 부터 주입받아서 사용하도록 변경

f-lab-dev commented on src/main/java/com/younghun/hibusgo/utils/LoginUtil.java:
```diff
@@ -1,53 +1,54 @@
-    return  (String) session.getAttribute(ACCOUNT_MEMBER_ID);
+  public String getLoginAccountId() {
+    String accountId = (String) session.getAttribute(ACCOUNT_MEMBER_ID);
+    return accountId != null ? accountId : null;
```
이런 경우는 `Optional`을 사용해서 null safe하게 사용할 수 있는 방법이 있습니다~
Optional에 대해 알아보시면 좋겠네요

0hun committed:
#3 로그인 기능

리뷰 반영

Optional 사용해서 null safe하게 소스 변경
